### PR TITLE
Update bundled WCSLIB to version 7.9

### LINKED ISSUE #12946: Cannot swapaxes when one of the WCS axis is WCS-TAB
@mcara has the file in the example. I am not sure if I can publicly share it.

```
Filename: Level2_det_image_seq1_MIRIFULONG_34LONGexp1_s3d.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     244   ()      
  1  SCI           1 ImageHDU        68   (47, 45, 1779)   float32   
  2  ERR           1 ImageHDU        12   (47, 45, 1779)   float32   
  3  DQ            1 ImageHDU        12   (47, 45, 1779)   int32 (rescales to uint32)   
  4  WMAP          1 ImageHDU        10   (47, 45, 1779)   float32   
  5  WCS-TABLE     1 BinTableHDU     14   1R x 1C   ['1779E']   
  6  ASDF          1 BinTableHDU     11   1R x 1C   [30659B]   
```
```
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN'  'DEC--TAN'  'WAVE-TAB'  
CRVAL : 0.000358857352731444  -4.1900237747446e-05  1.0  
CRPIX : 24.0  23.0  1.0  
PC1_1 PC1_2 PC1_3  : -1.0  0.0  0.0  
PC2_1 PC2_2 PC2_3  : 0.0  1.0  0.0  
PC3_1 PC3_2 PC3_3  : 0.0  0.0  1.0  
CDELT : 5.55555563833978e-05  5.55555563833978e-05  1.0  
NAXIS : 47  45  1779
```
```python
from astropy.io import fits
from astropy.wcs import WCS

lvl2_file = 'Level2_det_image_seq1_MIRIFULONG_34LONGexp1_s3d.fits'
pf = fits.open(lvl2_file)
w = WCS(pf['SCI'].header, pf)
w.swapaxes(0, 2)  # Error (traceback below)
w.wcs.sub([3, 2, 1])  # What it is really doing, same error
```
```
.../astropy/astropy/wcs/wcs.py in swapaxes(self, ax0, ax1)
   3054         inds[ax0], inds[ax1] = inds[ax1], inds[ax0]
   3055 
-> 3056         return self.sub([i+1 for i in inds])
   3057 
   3058     def reorient_celestial_first(self):

.../astropy/astropy/wcs/wcs.py in sub(self, axes)
    604 
    605         # Subset the WCS
--> 606         copy.wcs = copy.wcs.sub(axes)
    607         copy.naxis = copy.wcs.naxis
    608 

InconsistentAxisTypesError: ERROR 4 in wcs_types() at line 3074 of file cextern/wcslib/C/wcs.c:
Table parameters set for non-table axis type.
```

What I really want to do is actually this:

```python
from astropy import units as u
from astropy.io import fits
from astropy.wcs import WCS
from specutils import Spectrum1D

lvl2_file = 'Level2_det_image_seq1_MIRIFULONG_34LONGexp1_s3d.fits'
pf = fits.open(lvl2_file)
w = WCS(pf['SCI'].header, pf)
flux = pf['SCI'].data << u.Unit(pf['SCI'].header['BUNIT'])
sp = Spectrum1D(flux, wcs=w, mask=None)
```

But `specutils` tries to swap the axes, and hence WCS error as reported.

Is this just not possible (hence need a workaround at `specutils`) or a WCS bug?

[🐱](https://jira.stsci.edu/browse/JDAT-1547)

### PR DESCRIPTION
**Pull request by mcara**
### Description
Updates bundled WCSLIB to the latest release - 7.9 that includes several bug fixes, see https://www.atnf.csiro.au/people/mcalabre/WCS/CHANGES

Fixes #12946

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/wcs/setup_package.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from extension_helpers import import_file, write_if_different, get_compiler, pkg_config
 
 WCSROOT = os.path.relpath(os.path.dirname(__file__))
-WCSVERSION = "7.7"
```
Comment (pllim) [inline astropy/wcs/setup_package.py:19]: Huh, what happened to 7.8?

File: docs/changes/wcs/12994.feature.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Updated bundled ``WCSLIB`` to version 7.9 with several bugfixes. For a list of
```
Comment (pllim) [inline docs/changes/wcs/12994.feature.rst:1]:
```suggestion
Updated bundled WCSLIB to version 7.9 with several bugfixes. For a list of
```

**Comment by pllim** [comment]
Failure is related:

```
_____________________ TestWcsWithTime.test_keywods2wcsprm ______________________

self = <astropy.wcs.tests.test_wcs.TestWcsWithTime object at 0x7f27a5c152e0>

    def test_keywods2wcsprm(self):
        """ Make sure Wcsprm is populated correctly from the header."""
    
        ctype = [self.header[val] for val in self.header["CTYPE*"]]
        crval = [self.header[val] for val in self.header["CRVAL*"]]
        crpix = [self.header[val] for val in self.header["CRPIX*"]]
        cdelt = [self.header[val] for val in self.header["CDELT*"]]
        cunit = [self.header[val] for val in self.header["CUNIT*"]]
        assert list(self.w.wcs.ctype) == ctype
>       assert list(self.w.wcs.axis_types) == [2200, 2201, 3300, 0]
E       assert [2200, 2201, 3300, 4000] == [2200, 2201, 3300, 0]
E         At index 3 diff: 4000 != 0
E         Full diff:
E         - [2200, 2201, 3300, 0]
E         + [2200, 2201, 3300, 4000]
E         ?                    +++

../../astropy/wcs/tests/test_wcs.py:1392: AssertionError
```

File: docs/changes/wcs/12994.feature.rst (hunk 2 above)
Comment (mcara) [reply docs/changes/wcs/12994.feature.rst:1]: Why? in the past it was recorder as \``WCSLIB\``

Comment (pllim) [reply docs/changes/wcs/12994.feature.rst:1]: Was it? If so, then just ignore.

**Comment by pllim** [comment]
pyinstaller failure is also unrelated

**Comment by olebole** [comment]
As I am going to apply this partially to the Debian astropy package (which unbundles wcslib): Would it be possible in future to separate the upstream patch from the required local changes in astropy? This would my life easier, and may also be helpful to keep Mark Calabretta's changes separated from the self-authored ones.

**Comment by dhomeier** [comment]
If I am not mistaken all the upstream changes are inside `cextern/wcslib`, and the only ones that did not come with the WCSLIB source inside `astropy/wcs` (+ the changelog of course). Still could go into two commits.

**Comment by pllim** [comment]
@olebole , since you unbundle it, does `astropy/wcs/setup_package.py` matter to you? If not, I think only the test update might affect your patch.

**Comment by olebole** [comment]
@plim, you are right, it is just the test update. For this patch it was easy enough to find out (already in Debian unstable). Especially for wcslib, it is quite easy, as original files and astropy adjustments are in separate directories. However, I would still wish that the update of the original and the required adjustments are in separate commits, since then I would not need to separate them manually, and also I can be sure that the wcslib files were unchanged (so, no "quick wcslib fix" would slip in unnoticed). IMO this would make this type of PRs easier to understand.

**Comment by pllim** [comment]
@olebole , that is a fair request but I am not sure where to document such a policy.  @mcara or @saimn , what is the document that you follow to update bundled libs?

**Comment by mcara** [comment]
And I would like to understand better what is the issue here. Specifically, how would you split this particular PR into two parts? Unit test could have been updated in a separate PR *before* updating WCSLIB. But where (into which PR - first or second) would one put changes in the `astropy/wcs/setup_package.py`? What about changelog?

**Comment by dhomeier** [comment]
As I understood it it's just about putting the update `cextern/wcslib` proper into one commit, and the `astropy` ones into another, and perhaps more importantly, should any patches have been applied to `wcslib` itself, keep them in their own commit as well. I think that's the policy that has already been followed whenever the latter came up, such as #11260, since those should generally be reported upstream to Mark Calabretta for future releases anyway.
There is a rolling reminder in #9018 where these recommendations might just be added as comment or to the main description; not sure how effective that would be.

**Comment by mcara** [comment]
But I agree that any patches (emergency bug fixes) to the WCSLIB that we implement on our own (until upstream is updated) must be made in separate PRs.

**Comment by dhomeier** [comment]
You can still push all commits at once – no need to have a separate CI run for each. Not sure if a separate PR is useful for the patches – if they are fixing something really broken, the prior PR without them should not be merged!

**Comment by olebole** [comment]
I think that everything should still go into a single PR; also to show that it belongs together, and to test everything before it is committed in one go. Just make two commits instead of one.

**Comment by pllim** [comment]
> There is a rolling reminder in https://github.com/astropy/astropy/issues/9018 where these recommendations might just be added as comment or to the main description; not sure how effective that would be

Probably not very effective. People who do the release might not be the one who update these stuff. Maybe @mcara or @saimn can suggest a place they would intuitively look for such info, since they recently updated wcslib and cfitsio, respectively.

**Pull request merged**